metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java

```diff
@@ -0,0 +1,186 @@
+  private boolean wroteSomething;
+
+  /**
+   * The default constructor
+   * 
+   * @param config
+   */
+  public FileEmitter(FileEmitterConfig config) {
+
+    this.config = config;
+    this.eventFormatter = this.config.getEventFormatter();
+
+    DefaultPrettyPrinter pp = new DefaultPrettyPrinter()
+        .withObjectIndenter(new DefaultIndenter("    ", DefaultIndenter.SYS_LF))
+        .withArrayIndenter(new DefaultIndenter("    ", DefaultIndenter.SYS_LF));
+    this.dataTemplateCodec.setPrettyPrinter(pp);
+
+    try {
+      FileWriter fw = new FileWriter(config.getFileName(), false);
+      this.writter = new BufferedWriter(fw);
+      this.writter.append("[");
+      this.writter.newLine();
+    } catch (IOException e) {
+      throw new RuntimeException("Error while creating file", e);
+    }
+    this.wroteSomething = false;
+    log.debug("Emitter created successfully for " + this.config.getFileName());
+  }
+
+  @Override
+  public void close() throws IOException {
+    this.writter.newLine();
+    this.writter.append("]");
+    this.writter.close();
+    this.writter = null;
+    log.debug("Emitter closed for " + this.config.getFileName());
+  }
+
+  @Override
+  public Future<MetadataWriteResponse> emit(@SuppressWarnings("rawtypes") MetadataChangeProposalWrapper mcpw,
+      Callback callback) throws IOException {
+    return emit(this.eventFormatter.convert(mcpw), callback);
+  }
+
+  @Override
+  public Future<MetadataWriteResponse> emit(MetadataChangeProposal mcp, Callback callback) throws IOException {
+    if (this.writter == null) {
+      log.error("Emitter is closed. MCP could not be written on closed emitter.");
+      String errorMsg = "Emit function called after close.";
+      Future<MetadataWriteResponse> response = createFailFuture(errorMsg);
+      if (callback != null) {
+        callback.onFailure(new Exception(errorMsg));
+      }
+      return response;
+    }
+
+    String serializedMCP = dataTemplateCodec.mapToString(mcp.data());
+    if (wroteSomething) {
+      this.writter.append(",");
+      this.writter.newLine();
+    }
+    this.writter.append(serializedMCP);
+    wroteSomething = true;
+    log.debug("MCP written successfully: " + serializedMCP);
+    Future<MetadataWriteResponse> response = createSuccessFuture();
+    if (callback != null) {
+      try {
+        callback.onCompletion(response.get());
+      } catch (InterruptedException | ExecutionException e) {
+        log.warn("Callback could not be executed.", e);
+      }
+    }
+    return response;
+  }
+
+  @Override
+  public boolean testConnection() throws IOException, ExecutionException, InterruptedException {
+    throw new UnsupportedOperationException("testConnection not relavent for File Emitter");
+  }
+
+  @Override
+  public Future<MetadataWriteResponse> emit(List<UpsertAspectRequest> request, Callback callback) throws IOException {
+    throw new UnsupportedOperationException("UpsertAspectRequest cannot be sent to File");
```

Comment:
this error message is somewhat strange. Why not say "not relevant for File Emitter" as in testConnection() ?